### PR TITLE
githooks: create commit message template with prepare-commit-msg

### PR DESCRIPTION
**Pull request by VallariAg**
Resolves https://tracker.ceph.com/issues/51537

For files changes in `src/abc/xyz/file.ext`, `abcd/file2.txt`, and `file3.txt`, this would create a template with the file path (and remove the redundant "/src" part):
```
# Please enter the commit message for your changes.
# abc/xyz: <commit_message>
# abcd: <commit_message>
# file3.txt: <commit_message>
```
The developer can simply uncomment the most relevant lines and replace `<commit_message>` with their message text.  

To use git hooks, configure git to use `githooks` directory instead of `.git/hooks/` ([reason: maintaining git hooks in repository](https://stackoverflow.com/questions/427207/can-git-hook-scripts-be-managed-along-with-the-repository)) by running:
```
git config core.hooksPath githooks
```

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ideepika** [comment]
@VallariAg this is awesome, thanks for the contribution!

I was just curious why didn't you use .git/hooks directory itself? 

I tried your PR locally, we are not retaining default github comments, can we do a append to commit message defaults that github has, instead of complete template replacing?

**Comment by VallariAg** [comment]
Hey @ideepika, thank you!

> I was just curious why didn't you use .git/hooks directory itself?

Scripts inside`.git/hooks/` are not version controlled so they cannot be commited to the repo. So, I came across [this](https://stackoverflow.com/a/37861972) solution, to store these scripts in a directory which can be tracked by git. 
I think, we would need to add command `git config core.hooksPath githooks` to the installation script for these hooks to be used by everyone. 

> can we do a append to commit message defaults that github has, instead of complete template replacing?
> I'd like to see if possible would be Signed-off-by auto append,

Awesome, I'll make these changes! 

> We'd need to amend the commit in this PR as well with a git commit sign.

My bad, I'll edit my commit message with the changes you mentioned above.

**Comment by VallariAg** [comment]
Added the "signed-off-by" and the default commit message text.  
For changes on a file `/test.txt`, it creates the following commit message:
```
# Possible commit messages: 
# test.txt: <commit_message>

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# On branch test
# Changes to be committed:
#       new file:   test.txt
#
# Changes not staged for commit:
#       modified:   githooks/prepare-commit-msg
#       modified:   install-deps.sh
#
# Untracked files:
#       # Possible commit messages: 
#       gpg-key.asc
#       gpg-key.pub
#

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

```

**Comment by ideepika** [comment]
@VallariAg thanks for the update, added minor changes needs, looks good overall!

File: githooks/prepare-commit-msg [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#
+# An example hook script to prepare the commit log message.
+# Called by "git commit" with the name of the file that has the
+# commit message, followed by the description of the commit
+# message's source.  The hook's purpose is to edit the commit
+# message file.  If the hook fails with a non-zero status,
+# the commit is aborted.
+#
+# To enable this hook, rename this file to "prepare-commit-msg".
+
```
Comment (epuertat) [inline githooks/prepare-commit-msg:11]: It's always better [to write safer shell scripts](https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/):
```suggestion
set -euxo pipefail
```

**Review by epuertat — COMMENTED** [review]
Thanks a lot @VallariAg for your magnificent contribution here! 🥳 I left a few suggestions over there.

**Comment by epuertat** [comment]
@VallariAg regarding the file location, if you assume that users might have their own hooks in `.git/hooks` then the idea of changing the config is less feasible. So what I usually do here is creating symlinks from the hooks to files in the repo, and to make things more user-friendly you can provide a script (or document the command) for automating that.

File: githooks/prepare-commit-msg [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+if [[ -z "$COMMIT_SOURCE" ]]
+then
+  COMMIT_MSG_HEADERS="# Possible commit messages: "
+  for FILE in $(git diff --cached --name-only -r)
+  do
+    # According to files changed the commit message would start "<project_name>: <commit_message>"
+    # src/<project>/file.txt - "<project>:"
+    # file.txt               - "file.txt:"
+    # <project>/file.txt     - "<project>:" 
+    COMMIT_MSG_HEADERS+="\n$(echo "# $FILE: <commit_message>" | sed -E 's|src/+||g' | sed -r 's/(.*)(\/[a-z0-9.-]+)/\1 /')"
```
[the remark this answers is not in the document]
Comment (VallariAg) [reply githooks/prepare-commit-msg:29]: Creating a file for mapping paths to component sounds like a great idea! Yeah, keeping them in sync should not be a problem. 
Sounds awesome, doing this in python makes sense.

File: githooks/prepare-commit-msg [hunk 3]
```diff
@@ -9,33 +9,35 @@
...
     # <project>/file.txt     - "<project>:" 
-    echo "# $FILE: <commit_message>" | sed -E 's|src/+||g' | sed -r 's/(.*)(\/[a-z0-9.-]+)/\1 /' >> "$COMMIT_MSG_FILE"  
+    COMMIT_MSG_HEADERS+="\n$(echo "# $FILE: <commit_message>" | sed -E 's|src/+||g' | sed -r 's/(.*)(\/[a-z0-9.-]+)/\1 /')" 
   done
+  sed -i "1s;^;$COMMIT_MSG_HEADERS\n;" $COMMIT_MSG_FILE
 fi
 
+
+# Adds a Signed-off-by line to the message, that can still be edited (This is rarely a good idea).
+SOB=$(git var GIT_COMMITTER_IDENT | sed -n 's/^\(.*>\).*$/Signed-off-by: \1/p')
```
[the remark this answers is not in the document]
Comment (VallariAg) [reply githooks/prepare-commit-msg:36]: Oh, adding a complimentary hook to check the commit message is an awesome idea! 
I think we would be going for the `commit-msg` hook, as [they run in order](http://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks): `pre-commit` -> `prepare-commit-msg` ->  `commit-msg`.

Both hooks in python, got it!

File: githooks/prepare-commit-msg [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+# commit message, followed by the description of the commit
+# message's source.  The hook's purpose is to edit the commit
+# message file.  If the hook fails with a non-zero status,
+# the commit is aborted.
+#
+# To enable this hook, rename this file to "prepare-commit-msg".
+
+COMMIT_MSG_FILE=$1
+COMMIT_SOURCE=$2
+SHA1=$3
```
[the remark this answers is not in the document]
Comment (VallariAg) [reply githooks/prepare-commit-msg:14]: I haven't tried using commit templates, I'll try and test that!

**Comment by VallariAg** [comment]
> @VallariAg regarding the file location, if you assume that users might have their own hooks in `.git/hooks` then the idea of changing the config is less feasible. So what I usually do here is creating symlinks from the hooks to files in the repo, and to make things more user-friendly you can provide a script (or document the command) for automating that.

Oh, yes, that makes sense! symlinks it is, then.

**Comment by VallariAg** [comment]
I've pushed some of the work I did on this, using python scripts to achieve the suggested results. 

Though while working on this, I realized I need to re-approach a part of how commit-msgs are editing messages because of unexpected results in few cases (example: using "--amend" doesn't show the commit message template and use of these hooks with "-m" needs more customization).
I'll get back to this soon! This would really be a convenient feature to have. Sorry for the lack of activity on this one!
When it's ready for review, I'll change it back to open PR from draft PR.

**Pull request closed without merging**